### PR TITLE
fix: Escape characters that could break URL

### PR DESCRIPTION
**Pull request by EyitopeIO**
@fangfufu 

This PR fixes https://github.com/fangfufu/httpdirfs/issues/138 and https://github.com/fangfufu/httpdirfs/issues/127 . It also expands the scope of, and superseeds the closed https://github.com/fangfufu/httpdirfs/pull/137.

We correctly escape the URL everywhere else except in the base URL, which is where a character could be that would break the URL. Since the base URL could also include an IP address, or domain name, we consider that too. Actually, a local server was the first test case I had access to.

It made sense to me to escape the base URL once in function `Linksystem_init` rather than attempting to modify function `make_link_relative` (in the snippet below) which is already good at what it does. https://github.com/fangfufu/httpdirfs/blob/d6d4af0c8c4d2d505a53eba86f93ccd3109529f4/src/link.c#L1134-L1141 

**Test cases**

```
httpdirfs -f -d --cache https://download.opensuse.org/repositories/Cloud:/Images:/ foo
httpdirfs -d --cache https://old-releases.ubuntu.com/releases/23.04 foo
httpdirfs -d --single-file-mode https://download.iopsys.eu/iopsys/toolchain/crosstools-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32-Rel1.12-full.tar.bz2 foo
httpdirfs -f https://www.crawler-test.com/urls/url_with_spaces/URL%20with%20spaces foo
httpdirfs -d -f --cache "http://0.0.0.0:9000/SP ACE/G B" foo
httpdirfs -d -f "http://0.0.0.0:9000/SP ACE/G B" foo
httpdirfs -d -f --cache "http://0.0.0.0:9000/SP%20ACE/G%20B" foo
httpdirfs -d -f --cache --single-file-mode "http://0.0.0.0:9000/SP ACE/G B/code.deb" foo
```

### CODE REVIEW

**Comment by EyitopeIO** [comment]
If this doesn't pass your review, I'd just go eat ice cream. Skills issue.

**Comment by fangfufu** [comment]
It really doesn't. I am sorry. I don't understand why you need regex here. 

All you had to do was escaping/unescaping at the correct time when the cache files / fuse files get created.

You clearly know how to program, but I don't think the way you architect your current solution is suitable for this project.

File: src/link.c [hunk 1]
```diff
@@ -1177,3 +1181,89 @@ static void make_link_relative(const char *page_url, char *link_url)
...
+ * \brief Pattern matching strings for URLs entered by user.
+ * \details The order is important as we match from most specifiec to least.
+ * For example, 192.168.1.1:80 is more specific than 192.168.1.1, even though
+ * visiting the latter refers to the former by default.
+ */
+const char *const ip_patterns[] = {
+    "([0-9]{1,3}\\.){3}[0-9]{1,3}:[0-9]*",  // IPv4 with port
+    "([0-9]{1,3}\\.){3}[0-9]{1,3}",         // IPv4 without port
+    NULL,
+};
```
Comment (fangfufu) [inline src/link.c:1196]: Why is this needed?

File: src/link.c [hunk 2]
```diff
@@ -273,11 +275,12 @@ static LinkTable *single_LinkTable_new(const char *url)
     return linktbl;
 }
 
-LinkTable *LinkSystem_init(const char *url)
+LinkTable *LinkSystem_init(const char *f_url)
 {
     if (pthread_mutex_init(&link_lock, NULL)) {
         lprintf(error, "link_lock initialisation failed!\n");
     }
+    char *url = escape_full_url(f_url);
```
Comment (fangfufu) [inline src/link.c:283]: What if individual files have weird filename?

**Comment by fangfufu** [comment]
This is really not what I expected. I will not merge this as it is.

File: src/link.c (hunk 2 above)
Comment (EyitopeIO) [reply src/link.c:283]: The existing codebase (before this PR) correctly handles individual filenames and directories. Consider the directory of my server already mounted in `foo` below. This also works in single file mode.

```
eyi@Zeero:~/Projects/HTTPDIRFS$ ls foo/
 appointment20231116-11-zyn4je.ics   Arduino   dipfree_en64.exe   dipfree_en.exe   happ:55  'SP ACE'   财富
eyi@Zeero:~/Projects/HTTPDIRFS$ ls foo/SP\ ACE/
'G B'  'SHA SUMS'
eyi@Zeero:~/Projects/HTTPDIRFS$ ls -l foo/财富
total 0
-r--r--r-- 1 eyi eyi 20 Apr 16 20:00 yam.txt
eyi@Zeero:~/Projects/HTTPDIRFS$ 
``` 

Our main problem is that we do not escape invalid characters in the base URL. As [this line](https://github.com/fangfufu/httpdirfs/blob/d6d4af0c8c4d2d505a53eba86f93ccd3109529f4/src/link.c#L581) shows, we just begin downloading the base URL without first escaping.

The whole essence of this PR (a wider scope than the previous hack of the space character) is to just escape the base URL.

The reason for the regex is because the server I was testing with was running locally on my PC on port 9000, since I could not find public servers with all sorts of funny URLs and files. 

So one of my test command was `httpdirfs -d -f --cache "http://0.0.0.0:9000/财富" foo`  

I could remove the regex part, but we still need to escape the base URL in addition to the files and directories at that location.

Comment (fangfufu) [reply src/link.c:283]: No, what you suggested is unlikely the problem, because it works without the cache system enabled:
https://github.com/fangfufu/httpdirfs/issues/86#issuecomment-2045002052

**Comment by fangfufu** [comment]
The root cause of the problem is the mismatch between cache system and the link generation system. This pull request does not address that at all. 

It is likely due to stuff introduced here:
https://github.com/fangfufu/httpdirfs/blob/d6d4af0c8c4d2d505a53eba86f93ccd3109529f4/src/link.c#L469
